### PR TITLE
new accesslog

### PR DESCRIPTION
**Pull request by wizardlyk**

### CODE REVIEW

File: api/access-log-handler.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	objectLayer ObjectLayer
+}
+
+func (a accessLogHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var accessLogger = helper.AccessLogger
+
+	startTime := time.Now().UnixNano() / 1e6
+	a.handler.ServeHTTP(w, r)
+	finishTime := time.Now().UnixNano() / 1e6
+	var request_time = finishTime - startTime
```
Comment (cxt90730) [inline api/access-log-handler.go]: no meaning

File: api/api-response.go [hunk 2]
```diff
@@ -199,12 +199,18 @@ func GenerateMultiDeleteResponse(quiet bool, deletedObjects []ObjectIdentifier,
...
 
+var ResponseRecorderInstance *ResponseRecorder
 // WriteSuccessResponse write success headers and response if any.
 func WriteSuccessResponse(w http.ResponseWriter, response []byte) {
 	if response == nil {
 		w.WriteHeader(http.StatusOK)
 		return
 	}
+
+	ResponseRecorderInstance = NewResponseRecorder(w)
```
Comment (cxt90730) [inline api/api-response.go]: remove it

File: helper/config.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ type Config struct {
 	IamSecret                  string `toml:"iam_secret"`
 	IamVersion                 string `toml:"iam_version"`
 	LogPath                    string `toml:"log_path"`
+	AccessLogPath              string `toml:"access_log_path"`
+	Format              	   string `toml:"format"`
```
Comment (cxt90730) [inline helper/config.go]: AccessLogFormat

File: log/access-log.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package log
```
Comment (cxt90730) [inline log/access-log.go]: remove this file

File: main.go [hunk 5]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 var logger *log.Logger
+var accessLogger *log.AccessLogger
```
Comment (cxt90730) [inline main.go]: the type should be *log.Logger

File: api/access-log-handler.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	newReplacer := NewReplacer(r, ResponseRecorderInstance, "-")
+	format := newReplacer.Replace(helper.CONFIG.AccessLogFormat)
+	response := newReplacer.Replace(format)
+	helper.AccessLogger.Println(5, response)
+}
+
+func SetAccessLogHandler(handler http.Handler, objectLayer ObjectLayer) http.Handler {
+	return accessLogHandler{
+		handler:     handler,
+		objectLayer: objectLayer,
```
Comment (cxt90730) [inline api/access-log-handler.go]: no meaning

File: api/recorder.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package api
```
Comment (cxt90730) [inline api/recorder.go]: move to access log handler

File: api/recorder.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
```
Comment (cxt90730) [inline api/recorder.go]: change name

File: integrate/yigconf/yig.toml [hunk 9]
```diff
@@ -1,6 +1,8 @@
 s3domain = ["s3.test.com","s3-internal.test.com"]
 region = "cn-bj-1"
 log_path = "/var/log/yig/yig.log"
+access_log_path = "/var/log/yig/access.log"
+format = "{combined}"
```
Comment (cxt90730) [inline integrate/yigconf/yig.toml]: fix

File: main.go [hunk 10]
```diff
@@ -28,6 +29,16 @@ func main() {
 
 	helper.SetupConfig()
 
+	//access log
```
Comment (cxt90730) [inline main.go]: move to SetAccessHandler

File: api/access-log-handler.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	accessLogger.Println(5, response)
+}
+
+func SetAccessLogHandler(handler http.Handler, _ ObjectLayer) http.Handler {
+	//access log
+	a, err := os.OpenFile(helper.CONFIG.AccessLogPath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		panic("Failed to open access log file " + helper.CONFIG.AccessLogPath)
+	}
+	defer a.Close()
```
Comment (cxt90730) [inline api/access-log-handler.go]: close?

File: api/access-log-handler.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		status:                http.StatusOK,
+		start:                 time.Now(),
+	}
+}
+
+type accessLogHandler struct {
+	handler http.Handler
+}
+
+var ResponseRecorderInstance *ResponseRecorder
```
Comment (thesues) [inline api/access-log-handler.go]: What's the usage of this GLOBAL variable?

File: api/generic-handlers.go [hunk 13]
```diff
@@ -26,6 +27,12 @@ import (
 	"github.com/journeymidnight/yig/signature"
 )
 
+type MiddleWare interface {
```
Comment (thesues) [inline api/generic-handlers.go]: Who uses this MiddleWare?

File: api/generic-handlers.go [hunk 14]
```diff
@@ -16,6 +16,7 @@
 
 package api
 
+import "C"
```
Comment (thesues) [inline api/generic-handlers.go]: who uses this package?

File: api/generic-handlers.go (hunk 13 above)
Comment (wizardlyk) [reply api/generic-handlers.go]: accessLogHandler, logHandler and the four handlers in generic-handlers

File: api/log-replacer.go [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+// getSubstitution retrieves value from corresponding key
+func (r *replacer) getSubstitution(key string) string {
+	// search custom replacements first
+	if value, ok := r.customReplacements[key]; ok {
+		return value
+	}
+
+	// search request headers then
+	if key[1] == '>' {
```
Comment (cannium) [inline api/log-replacer.go]: What features does your small "replace language" support?

File: api/log-replacer.go [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+		// strip out scanned parts
+		s = s[idxEnd+1:]
+	}
+
+	// append unscanned parts
+	return result + unescapeBraces(s)
+}
+
+// getPeerCert returns peer certificate
+func (r *replacer) getPeerCert() *x509.Certificate {
```
Comment (cannium) [inline api/log-replacer.go]: What's the purpose of this method?

File: api/generic-handlers.go [hunk 17]
```diff
@@ -38,64 +38,33 @@ func RegisterHandlers(router *mux.Router, objectLayer ObjectLayer, handlerFns ..
 	return f
 }
 
+func CloseHandlers()  {
+	helper.Logger.Println(5, "Stopping API handlers...")
+	AccessLogHandler.Close(AccessLogHandler{})
```
Comment (cannium) [inline api/generic-handlers.go]: This is a type conversion, not a method call.

**Pull request merged**